### PR TITLE
Use packed stack-based version of  Slot0 structure

### PR DESCRIPTION
**Pull request by 0xVolosnikov**
This pull request does not relate to any issue. In my opinion, it is more convenient to discuss this change if you have an already written implementation in a pull request.

## Description of changes

The slot0 structure by design fits into one storage slot. However, when using the memory version of this structure, each field occupies a separate memory slot. In addition, the memory structure takes up more space on the solidity "stack".

The changes I propose allows to read this structure onto the stack and operate with the fields of this structure directly on the stack. As a result, the amount of memory used is reduced, which a little bit reduces gas costs and bytecode size.

Since the changes are more or less significant, I would first like to hear the opinion of the community and maintainers regarding the proposal to use a structure on the stack before adding tests and so on.

### CODE REVIEW

File: src/libraries/Pool.sol [hunk 1]
```diff
@@ -105,9 +92,15 @@ library Pool {
...
-        if (tickUpper > TickMath.MAX_TICK) revert TickUpperOutOfBounds(tickUpper);
+        if (tickLower >= tickUpper) {
+            revert TicksMisordered(tickLower, tickUpper);
+        }
+        if (tickLower < TickMath.MIN_TICK) {
+            revert TickLowerOutOfBounds(tickLower);
+        }
+        if (tickUpper > TickMath.MAX_TICK) {
+            revert TickUpperOutOfBounds(tickUpper);
+        }
```
Comment (shuhuiluo) [inline src/libraries/Pool.sol:103]: Unrelated code style changes.

File: src/libraries/Pool.sol [hunk 2]
```diff
@@ -443,10 +441,12 @@ library Pool {
...
         }
 
-        (self.slot0.sqrtPriceX96, self.slot0.tick) = (state.sqrtPriceX96, state.tick);
+        self.slot0.storePacked(slot0Start.setTick(state.tick).setSqrtPriceX96(state.sqrtPriceX96));
 
         // update liquidity if it changed
-        if (cache.liquidityStart != state.liquidity) self.liquidity = state.liquidity;
+        if (cache.liquidityStart != state.liquidity) {
+            self.liquidity = state.liquidity;
+        }
```
Comment (shuhuiluo) [inline src/libraries/Pool.sol:449]: Unrelated code style change.

File: src/types/Slot0.sol [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            _packed := sload(pointer.slot)
+        }
+    }
+}
+
+/// @dev This library is used to get/set values in packed structure on stack
+library Slot0PackedLibrary {
+    uint256 private constant UINT160_MASK = 0x00FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
+    uint256 private constant INT24_MASK = 0xFFFFFF;
+    uint256 private constant UINT24_MASK = 0xFFFFFF;
```
Comment (shuhuiluo) [inline src/types/Slot0.sol:44]:
```suggestion
    uint160 private constant UINT160_MASK = 0x00FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
    uint24 private constant INT24_MASK = 0xFFFFFF;
    uint24 private constant UINT24_MASK = 0xFFFFFF;
```

File: src/types/Slot0.sol [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    uint256 private constant TICK_OFFSET = 160;
+    uint256 private constant PROTOCOL_FEE_OFFSET = 184;
+    uint256 private constant LP_FEE_OFFSET = 208;
```
Comment (shuhuiluo) [inline src/types/Slot0.sol:48]:
```suggestion
    uint8 private constant TICK_OFFSET = 160;
    uint8 private constant PROTOCOL_FEE_OFFSET = 184;
    uint8 private constant LP_FEE_OFFSET = 208;
```

File: src/types/Slot0.sol [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        /// @solidity memory-safe-assembly
+        assembly {
+            _sqrtPriceX96 := and(UINT160_MASK, _packed)
+        }
+    }
+
+    function tick(Slot0Packed _packed) internal pure returns (int24 _tick) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _tick := and(INT24_MASK, shr(TICK_OFFSET, _packed))
```
Comment (shuhuiluo) [inline src/types/Slot0.sol:61]:
```suggestion
            _tick := signextend(2, shr(TICK_OFFSET, _packed))
```
Makes it two's complement.

File: src/types/Slot0.sol [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        /// @solidity memory-safe-assembly
+        assembly {
+            _protocolFee := and(UINT24_MASK, shr(PROTOCOL_FEE_OFFSET, _packed))
+        }
+    }
+
+    function lpFee(Slot0Packed _packed) internal pure returns (uint24 _lpFee) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _lpFee := and(UINT24_MASK, shr(LP_FEE_OFFSET, _packed))
```
Comment (shuhuiluo) [inline src/types/Slot0.sol:75]:
```suggestion
            _lpFee := shr(LP_FEE_OFFSET, _packed)
```

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

File: src/PoolManager.sol [hunk 7]
```diff
@@ -114,8 +115,12 @@ contract PoolManager is IPoolManager, ProtocolFees, NoDelegateCall, ERC6909Claim
...
         if (key.tickSpacing > MAX_TICK_SPACING) revert TickSpacingTooLarge();
         if (key.tickSpacing < MIN_TICK_SPACING) revert TickSpacingTooSmall();
-        if (key.currency0 >= key.currency1) revert CurrenciesOutOfOrderOrEqual();
-        if (!key.hooks.isValidHookAddress(key.fee)) revert Hooks.HookAddressNotValid(address(key.hooks));
+        if (key.currency0 >= key.currency1) {
+            revert CurrenciesOutOfOrderOrEqual();
+        }
+        if (!key.hooks.isValidHookAddress(key.fee)) {
+            revert Hooks.HookAddressNotValid(address(key.hooks));
+        }
```
[the remark this answers is not in the document]
Comment (hensha256) [reply src/PoolManager.sol:123]: agreed - we use foundry.toml's default line length

File: src/PoolManager.sol [hunk 8]
```diff
@@ -59,9 +60,9 @@ contract PoolManager is IPoolManager, ProtocolFees, NoDelegateCall, ERC6909Claim
         override
         returns (uint160 sqrtPriceX96, int24 tick, uint24 protocolFee, uint24 lpFee)
     {
-        Pool.Slot0 memory slot0 = pools[id].slot0;
+        Slot0Packed slot0 = pools[id].slot0.loadPacked();
```
[the remark this answers is not in the document]
Comment (snreynolds) [reply src/PoolManager.sol:63]: I see that Slot0 is still used - but why? If storage is already packed, why do we need to manually pack. 

Id prefer if we could write this such that we match how storage is packed and so we only really need to have the type

Slot0 is bytes32;

**Comment by snreynolds** [comment]
Btw in general, I do really like this change and it would be great to get tests I think! cc @hensha256

File: src/PoolManager.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (0xVolosnikov) [reply src/PoolManager.sol:170]: > yeah you can run `forge fmt`

Doesn't help. Maybe I'm missing some additional config?

File: src/PoolManager.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (0xVolosnikov) [reply src/PoolManager.sol]: Built-in structure is somewhat easier for fields change process. So I tried to keep it. 
Anyway, I removed memory-based structure completely.

File: src/types/Slot0.sol (hunk 6 above)
Comment (0xVolosnikov) [reply src/types/Slot0.sol:58]: For some reason gas increases

File: src/types/Slot0.sol (hunk 3 above)
Comment (0xVolosnikov) [reply src/types/Slot0.sol]: No effects

File: src/types/Slot0.sol (hunk 4 above)
Comment (shuhuiluo) [reply src/types/Slot0.sol]: Great. That means solc is smart enough to use `PUSH1` for these instead of `PUSH32` so bytecode size is already optimized!

File: src/types/Slot0.sol [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    // the current tick
+    int24 tick;
+    // protocol fee, expressed in hundredths of a bip
+    // upper 12 bits are for 1->0, and the lower 12 are for 0->1
+    // the maximum is 1000 - meaning the maximum protocol fee is 0.1%
+    // the protocolFee is taken from the input first, then the lpFee is taken from the remaining input
+    uint24 protocolFee;
+    // used for the lp fee, either static at initialize or dynamic via hook
+    uint24 lpFee;
+*/
```
Comment (snreynolds) [inline src/types/Slot0.sol:18]: Could you maybe just also add a comment that we dont use a struct so that we get the benefits of a packed structure when we reference slot0 in memory/stack

File: src/types/Slot0.sol [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+type Slot0 is bytes32;
+
+using Slot0Library for Slot0 global;
+
+library Slot0Library {
+    uint160 private constant UINT160_MASK = 0x00FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
```
Comment (snreynolds) [inline src/types/Slot0.sol:25]: why is this not just `0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF` ?

File: src/types/Slot0.sol [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    function tick(Slot0 _packed) internal pure returns (int24 _tick) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _tick := signextend(2, shr(TICK_OFFSET, _packed))
+        }
+    }
+
+    function protocolFee(Slot0 _packed) internal pure returns (uint24 _protocolFee) {
```
Comment (snreynolds) [inline src/types/Slot0.sol:48]: I wonder now that we are writing getters manually if we can just get rid of the getters for 0 to 1 and 1 to 0 in ProtocolFeeLibrary and just do it directly in this library? Maybe would save some overhead - I still like having a ProtcocolFeeLibrary but worth playing around if we can remove getZeroForOneFee / getOneForZeroFee

File: src/types/Slot0.sol (hunk 10 above)
Comment (0xVolosnikov) [reply src/types/Slot0.sol:25]: Compiler treats it as an address literal and is really angry because of incorrectly formatted address

File: src/types/Slot0.sol (hunk 11 above)
Comment (0xVolosnikov) [reply src/types/Slot0.sol:48]: Sounds reasonable! Then we can remove `protocolFee` completely and treat it as two packed `uint12` inside of a structure.

In this case, it is logical to change the public getter for slot0 too

File: src/libraries/Pool.sol [hunk 12]
```diff
@@ -111,23 +98,28 @@ library Pool {
...
 
         tick = TickMath.getTickAtSqrtPrice(sqrtPriceX96);
 
-        self.slot0 = Slot0({sqrtPriceX96: sqrtPriceX96, tick: tick, protocolFee: protocolFee, lpFee: lpFee});
+        self.slot0 = 
+            Slot0.wrap(bytes32(0))
+                .setSqrtPriceX96(sqrtPriceX96)
+                .setTick(tick)
+                .setProtocolFee(protocolFee)
+                .setLpFee(lpFee);
```
Comment (shuhuiluo) [inline src/libraries/Pool.sol:110]: How about a `pack` method if putting all components together? Should be more efficient.

Comment (0xVolosnikov) [reply src/libraries/Pool.sol:110]: I'm afraid that this will lead to a somewhat significant increase in bytecode size for the sake of a tiny gain in a function that is called once. In my opinion, it's better to have smaller bytecode, but more aggressive optimizer settings

Comment (0xVolosnikov) [reply src/libraries/Pool.sol:110]: I'll check what it does anyway

Comment (0xVolosnikov) [reply src/libraries/Pool.sol:110]: @shuhuiluo I tried to implement it as:

```solidity
  function pack(uint160 _sqrtPriceX96, int24 _tick, uint24 _protocolFee, uint24 _lpFee) internal pure returns (Slot0 _result) {
        /// @solidity memory-safe-assembly
        assembly {
          _result := and(UINT160_MASK, _sqrtPriceX96)
          _result := or(_result, shl(TICK_OFFSET, and(INT24_MASK, _tick)))
          _result := or(_result, shl(PROTOCOL_FEE_OFFSET, and(UINT24_MASK, _protocolFee)))
          _result := or(_result, shl(LP_FEE_OFFSET, and(UINT24_MASK, _lpFee)))
      }      
  }
```

and I see no changes in snapshots at all...

File: src/types/Slot0.sol (hunk 9 above)
Comment (0xVolosnikov) [reply src/types/Slot0.sol]: Added

File: src/libraries/Pool.sol (hunk 12 above)
Comment (shuhuiluo) [reply src/libraries/Pool.sol:110]: Interesting. The compiler may be able to reduce the math and the masking on `0`. Still, appreciate it.

File: src/types/Slot0.sol (hunk 11 above)
Comment (0xVolosnikov) [reply src/types/Slot0.sol:51]: I tried to implement `protocolFeeZeroForOne`/`protocolFeeOneForZero` as `uint12` in `Slot0` and do not see any gains in snapshots, looks like the optimizer itself substitutes and optimizes getters.
Probably such a change only makes sense to simplify the code, but imo the current version is already quite good.

Comment (shuhuiluo) [reply src/types/Slot0.sol:51]: Simple getters without side effects should be automatically inlined by the compiler.

File: src/types/Slot0.sol (hunk 10 above)
Comment (hensha256) [reply src/types/Slot0.sol:28]: could just change it to `type(uint160).max`? That feels more reliable anyway as you dont need to check you have the right number of Fs

File: src/types/Slot0.sol [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+ * int24 tick;
+ * 
+ * Protocol fee, expressed in hundredths of a bip, upper 12 bits are for 1->0, and the lower 12 are for 0->1
+ * the maximum is 1000 - meaning the maximum protocol fee is 0.1%
+ * the protocolFee is taken from the input first, then the lpFee is taken from the remaining input
+ * uint24 protocolFee;
+ * 
+ * Used for the lp fee, either static at initialize or dynamic via hook
+ * uint24 lpFee;
+ */
```
Comment (hensha256) [inline src/types/Slot0.sol:22]: imo worth writing in this comment about the unused bits. How many, where they are etc.

File: src/types/Slot0.sol (hunk 10 above)
Comment (0xVolosnikov) [reply src/types/Slot0.sol:28]: > could just change it to `type(uint160).max`?

Only direct number constants can be used in assembly. We can replace it with `type(uin160).max` but it will be a bit messier

File: src/types/Slot0.sol (hunk 13 above)
Comment (hensha256) [reply src/types/Slot0.sol:22]: Even a single line like

`24 bits empty | 24 bits lpFee | 24 bits protocolFee | 24 bits tick | 160 bits sqrtPriceX96`

i think would make this super clear what the layout is and the direction of the packing

Comment (0xVolosnikov) [reply src/types/Slot0.sol:25]: Added

File: src/types/Slot0.sol [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+            _result := or(and(not(shl(TICK_OFFSET, INT24_MASK)), _packed), shl(TICK_OFFSET, and(INT24_MASK, _tick)))
+        }
+    }
+
+    function setProtocolFee(Slot0 _packed, uint24 _protocolFee) internal pure returns (Slot0 _result) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _result :=
+                or(
+                    and(not(shl(PROTOCOL_FEE_OFFSET, UINT24_MASK)), _packed),
```
Comment (hensha256) [inline src/types/Slot0.sol:88]: presumably optimiser is catching that `not(shl(PROTOCOL_FEE_OFFSET, UINT24_MASK))`  is always the same?

File: src/types/Slot0.sol (hunk 10 above)
Comment (hensha256) [reply src/types/Slot0.sol:31]: oh yeah damn, annoying

File: src/types/Slot0.sol (hunk 14 above)
Comment (0xVolosnikov) [reply src/types/Slot0.sol:88]: Will check it

Comment (0xVolosnikov) [reply src/types/Slot0.sol:88]: Indeed, attempts to use constants instead of `not(shl(...))` do not change anything or even make gas worse

File: src/types/Slot0.sol [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+ * Used for the lp fee, either static at initialize or dynamic via hook
+ * uint24 lpFee;
+ */
+type Slot0 is bytes32;
+
+using Slot0Library for Slot0 global;
+
+library Slot0Library {
+    uint160 internal constant UINT160_MASK = 0x00FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
+    uint24 internal constant INT24_MASK = 0xFFFFFF;
```
Comment (0xVolosnikov) [inline src/types/Slot0.sol:32]: `INT24_MASK` can be removed since it is equal to `UINT24_MASK`

**Comment by 0xVolosnikov** [comment]
Added tests

File: test/types/Slot0.t.sol [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+import {Test} from "forge-std/Test.sol";
+import {Slot0, Slot0Library} from "../../src/types/Slot0.sol";
+
+contract TestSlot0 is Test {
+    function test_slot0_constants_masks() public pure {
+        assertEq(Slot0Library.UINT160_MASK, type(uint160).max);
+        assertEq(Slot0Library.UINT24_MASK, type(uint24).max);
+    }
+
+    function test_slot0_fuzz_pack_unpack(uint160 sqrtPriceX96, int24 tick, uint24 protocolFee, uint24 lpFee) public pure {
```
Comment (hensha256) [inline test/types/Slot0.t.sol]: a nit to be in keeping with the repo

```suggestion
    function test_fuzz_slot0_pack_unpack(uint160 sqrtPriceX96, int24 tick, uint24 protocolFee, uint24 lpFee) public pure {
```

**Comment by hensha256** [comment]
@snreynolds anything else you wanna see before we merge?

File: src/types/Slot0.sol [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+ * Used for the lp fee, either static at initialize or dynamic via hook
+ * uint24 lpFee;
+ */
+type Slot0 is bytes32;
+
+using Slot0Library for Slot0 global;
+
+library Slot0Library {
+    uint160 internal constant UINT160_MASK = 0x00FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
+    uint24 internal constant UINT24_MASK = 0xFFFFFF;
```
Comment (hensha256) [inline src/types/Slot0.sol:33]: nit: would rather this was called `24_BIT_MASK` or something like that now that its used for both int24 and uint24

Comment (0xVolosnikov) [reply src/types/Slot0.sol]: Changed to `MASK_24_BITS`

**Review by hensha256 — COMMENTED** [review]
1 final nit, plus you need to handle merge conflicts (sorry we're merging lots today!). Then we can merge!

File: src/PoolManager.sol [hunk 18]
```diff
@@ -18,6 +18,7 @@ import {ERC6909Claims} from "./ERC6909Claims.sol";
 import {PoolId, PoolIdLibrary} from "./types/PoolId.sol";
 import {BalanceDelta, BalanceDeltaLibrary, toBalanceDelta} from "./types/BalanceDelta.sol";
 import {BeforeSwapDelta} from "./types/BeforeSwapDelta.sol";
+import {Slot0} from "./types/Slot0.sol";
```
[the remark this answers is not in the document]
Comment (0xVolosnikov) [reply src/PoolManager.sol]: Removed

File: src/libraries/Pool.sol [hunk 19]
```diff
@@ -297,25 +286,28 @@ library Pool {
...
-        state.tick = slot0Start.tick;
+        state.sqrtPriceX96 = slot0Start.sqrtPriceX96();
+        state.tick = slot0Start.tick();
         state.feeGrowthGlobalX128 = zeroForOne ? self.feeGrowthGlobal0X128 : self.feeGrowthGlobal1X128;
         state.liquidity = liquidityStart;
 
         // if the beforeSwap hook returned a valid fee override, use that as the LP fee, otherwise load from storage
-        slot0Start.lpFee =
-            params.lpFeeOverride.isOverride() ? params.lpFeeOverride.removeOverrideAndValidate() : slot0Start.lpFee;
+        {
```
Comment (hensha256) [inline src/libraries/Pool.sol:304]: nice thank you!

**Review by hensha256 — APPROVED** [review]
